File: web/src/types/todo.ts
/** Shared TypeScript types for Todo entities, matching backend Pydantic schemas. */

/** A todo item as returned by the API. */
export interface Todo {
  id: number;
  title: string;
  description: string | null;
  is_completed: boolean;
  created_at: string;
  updated_at: string;
}

/** Payload for creating a new todo. */
export interface TodoCreate {
  title: string;
  description?: string | null;
}

/** Payload for updating an existing todo (all fields optional). */
export interface TodoUpdate {
  title?: string;
  description?: string | null;
  is_completed?: boolean;
}

File: api/app/schemas/todo.py
"""Pydantic v2 schemas for Todo request/response DTOs."""

import datetime

from pydantic import BaseModel, ConfigDict, Field


class TodoCreate(BaseModel):
    """Schema for creating a new todo."""

    title: str = Field(..., min_length=1, max_length=200, description="Todo title")
    description: str | None = Field(None, description="Optional detailed description")


class TodoUpdate(BaseModel):
    """Schema for updating an existing todo. All fields optional."""

    title: str | None = Field(None, min_length=1, max_length=200, description="Todo title")
    description: str | None = Field(None, description="Optional detailed description")
    is_completed: bool | None = Field(None, description="Completion status")


class TodoResponse(BaseModel):
    """Schema for returning a todo to the client."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    title: str
    description: str | None
    is_completed: bool
    created_at: datetime.datetime
    updated_at: datetime.datetime
